File: cli/internal/adapters/secretscrypto/secretscrypto.go
// Package secretscrypto — .cxtsecrets end-to-end encryption (WebCrypto and byte-compatible).
//
// Design (security review points):
//   - Server stores only ciphertext (E2E). Passwords are shared externally by the team (server not involved).
//   - KDF: PBKDF2-HMAC-SHA256, 600,000 iterations (OWASP 2023 recommendation), 16B random salt.
//   - AEAD: AES-256-GCM, 12B random nonce, AAD = "cxtsecrets:v1:<repoID>" —
//     Prevents key reuse across different repos (context binding).
//   - Quantum resistance: No asymmetric exchange, no Shor surface, AES-256 considered with Grover, 128-bit effective security (NIST current quantum resistance standard). Pure stdlib implementation.
package secretscrypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"regexp"
	"time"
)

// Iterations is the PBKDF2 iteration count (must be the same as Web).
const Iterations = 600_000

// Envelope is the ciphertext envelope stored on the server (common JSON format for Web/CLI).
type Envelope struct {
	Version    int    `json:"version"` // 1
	KDF        string `json:"kdf"`     // "PBKDF2-SHA256"
	Iterations int    `json:"iterations"`
	SaltB64    string `json:"salt_b64"`
	Cipher     string `json:"cipher"` // "AES-256-GCM"
	NonceB64   string `json:"nonce_b64"`
	CipherB64  string `json:"ciphertext_b64"`
	// Fingerprint is the password fingerprint (first 12 hex) — validation value to check if the team is using the same password. Does not expose password characters (see Fingerprint below).
	Fingerprint string    `json:"fingerprint,omitempty"`
	UpdatedAt   time.Time `json:"updated_at,omitempty"`
	UpdatedBy   string    `json:"updated_by,omitempty"`
}

// fpSalt is a fixed salt for fingerprints (repo-bound) — deterministic unlike the encryption random salt, always produces the same fingerprint for the same (password, repo).
func fpSalt(repoID string) []byte { return []byte("cxtsecrets-fp:v1:" + repoID) }

// Fingerprint is the password's fingerprint (SHA-256(PBKDF2(pass, fixed salt, 600k)) first 12 hex). KDF cost is the same as key derivation, preventing new surfaces in offline attacks while not exposing the password plaintext for "same password" comparison. Same as Web and byte compatibility.
func Fingerprint(passphrase, repoID string) string {
	sum := sha256.Sum256(pbkdf2Key([]byte(passphrase), fpSalt(repoID), Iterations, 32))
	return hex.EncodeToString(sum[:])[:12]
}

// rePassphrase — 4 or more words, single space separation (no leading/trailing spaces, consecutive spaces, punctuation, or digits allowed). Same as Web passphrase.ts.
var rePassphrase = regexp.MustCompile(`^[a-zA-Z]+( [a-zA-Z]+){3,}$`)

// ValidatePassphrase checks team password format (must match rules): 4 or more words, 12 characters or more. ASCII English characters only, so len(bytes) == len(characters) for accurate length check.
func ValidatePassphrase(p string) error {
	if len(p) < 12 || !rePassphrase.MatchString(p) {
		return errors.New("Team passphrase must be at least 4 English words and 12 characters (e.g., harbor twist ledger cousin)")
	}
	return nil
}

// pbkdf2 (RFC 8018, HMAC-SHA256) — no external dependencies, using stdlib.
func pbkdf2Key(password, salt []byte, iter, keyLen int) []byte {
	prf := func(data []byte) []byte {
		h := hmac.New(sha256.New, password)
		h.Write(data)
		return h.Sum(nil)
	}
	numBlocks := (keyLen + sha256.Size - 1) / sha256.Size
	out := make([]byte, 0, numBlocks*sha256.Size)
	for block := 1; block <= numBlocks; block++ {
		idx := make([]byte, 4)
		binary.BigEndian.PutUint32(idx, uint32(block))
		u := prf(append(append([]byte{}, salt...), idx...))
		t := append([]byte{}, u...)
		for i := 1; i < iter; i++ {
			u = prf(u)
			for j := range t {
				t[j] ^= u[j]
			}
		}
		out = append(out, t...)
	}
	return out[:keyLen]
}

func aad(repoID string) []byte { return []byte("cxtsecrets:v1:" + repoID) }

// Encrypt takes a plaintext and a passphrase to create an envelope.
func Encrypt(passphrase, plaintext, repoID string) (Envelope, error) {
	salt := make([]byte, 16)
	nonce := make([]byte, 12)
	if _, err := rand.Read(salt); err != nil {
		return Envelope{}, err
	}
	if _, err := rand.Read(nonce); err != nil {
		return Envelope{}, err
	}
	key := pbkdf2Key([]byte(passphrase), salt, Iterations, 32)
	block, err := aes.NewCipher(key)
	if err != nil {
		return Envelope{}, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return Envelope{}, err
	}
	ct := gcm.Seal(nil, nonce, []byte(plaintext), aad(repoID))
	return Envelope{
		Version: 1, KDF: "PBKDF2-SHA256", Iterations: Iterations,
		SaltB64: base64.StdEncoding.EncodeToString(salt),
		Cipher:  "AES-256-GCM", NonceB64: base64.StdEncoding.EncodeToString(nonce),
		CipherB64:   base64.StdEncoding.EncodeToString(ct),
		Fingerprint: Fingerprint(passphrase, repoID),
	}, nil
}

// ValidateEnvelope checks the wire envelope contract before starting an expensive KDF.
func ValidateEnvelope(env Envelope) error {
	if env.Version != 1 || env.KDF != "PBKDF2-SHA256" || env.Cipher != "AES-256-GCM" {
		return errors.New("Unsupported envelope format")
	}
	if env.Iterations != Iterations {
		return fmt.Errorf("Unsupported KDF iteration count: %d", env.Iterations)
	}
	salt, err := base64.StdEncoding.DecodeString(env.SaltB64)
	if err != nil || len(salt) != 16 {
		return errors.New("Envelope salt must be 16 bytes")
	}
	nonce, err := base64.StdEncoding.DecodeString(env.NonceB64)
	if err != nil || len(nonce) != 12 {
		return errors.New("Envelope nonce must be 12 bytes")
	}
	ciphertext, err := base64.StdEncoding.DecodeString(env.CipherB64)
	if err != nil || len(ciphertext) < 16 {
		return errors.New("AES-GCM ciphertext format is invalid")
	}
	if env.Fingerprint != "" {
		if len(env.Fingerprint) != 12 {
			return errors.New("envelope fingerprint format is invalid")
		}
		for _, r := range env.Fingerprint {
			if (r < '0' || r > '9') && (r < 'a' || r > 'f') {
				return errors.New("envelope fingerprint format is invalid")
			}
		}
	}
	return nil
}

// Decrypt decrypts the envelope. A passphrase mismatch or corruption results in an error (GCM authentication failure).
func Decrypt(passphrase string, env Envelope, repoID string) (string, error) {
	if err := ValidateEnvelope(env); err != nil {
		return "", err
	}
	salt, err := base64.StdEncoding.DecodeString(env.SaltB64)
	if err != nil {
		return "", err
	}
	nonce, err := base64.StdEncoding.DecodeString(env.NonceB64)
	if err != nil {
		return "", err
	}
	ct, err := base64.StdEncoding.DecodeString(env.CipherB64)
	if err != nil {
		return "", err
	}
	key := pbkdf2Key([]byte(passphrase), salt, env.Iterations, 32)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	pt, err := gcm.Open(nil, nonce, ct, aad(repoID))
	if err != nil {
		return "", errors.New("decryption failed — passphrase is incorrect or data is corrupted")
	}
	return string(pt), nil
}

File: frontend/web/src/secretscrypto.ts
// .cxtsecrets inter-machine encryption — CLI(Go secretscrypto) and byte-compatible.
// PBKDF2-SHA256(600k) + AES-256-GCM, AAD = "cxtsecrets:v1:<repoID>".
// Password/plaintext/key does not leave the browser — server only handles this envelope(ciphertext).
import type { MsgKey, Vars } from './i18n';

export interface SecretsEnvelope {
  version: number;
  kdf: string;
  iterations: number;
  salt_b64: string;
  cipher: string;
  nonce_b64: string;
  ciphertext_b64: string;
/** Password fingerprint (first 12 hex) — for comparison/blocking, plaintext not exposed. */
  fingerprint?: string;
  updated_at?: string;
  updated_by?: string;
}

export const ITERATIONS = 600_000;

const enc = new TextEncoder();
const dec = new TextDecoder();
const b64 = (buf: ArrayBuffer | Uint8Array) => btoa(String.fromCharCode(...new Uint8Array(buf)));
const unb64 = (s: string) => Uint8Array.from(atob(s), (c) => c.charCodeAt(0));

async function deriveKey(passphrase: string, salt: Uint8Array, iterations: number): Promise<CryptoKey> {
  const base = await crypto.subtle.importKey('raw', enc.encode(passphrase), 'PBKDF2', false, ['deriveKey']);
  return crypto.subtle.deriveKey(
    { name: 'PBKDF2', hash: 'SHA-256', salt: salt as BufferSource, iterations },
    base,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );
}

const aad = (repoId: string) => enc.encode(`cxtsecrets:v1:${repoId}`);
const hexstr = (buf: ArrayBuffer) => [...new Uint8Array(buf)].map((b) => b.toString(16).padStart(2, '0')).join('');

function validEnvelope(env: SecretsEnvelope): boolean {
  if (
    env.version !== 1 ||
    env.kdf !== 'PBKDF2-SHA256' ||
    env.cipher !== 'AES-256-GCM' ||
    env.iterations !== ITERATIONS
  ) {
    return false;
  }
  try {
    if (unb64(env.salt_b64).byteLength !== 16) return false;
    if (unb64(env.nonce_b64).byteLength !== 12) return false;
    if (unb64(env.ciphertext_b64).byteLength < 16) return false;
  } catch {
    return false;
  }
  return !env.fingerprint || /^[0-9a-f]{12}$/.test(env.fingerprint);
}

// fingerprint — password fingerprint(SHA-256(PBKDF2(pass, fixed salt, 600k)) first 12 hex).
// Fixed salt(repo binding) is deterministic, KDF cost is the same as encryption, not increasing attack surface,
// and does not expose the password plaintext. Go(secretscrypto.Fingerprint) and byte-compatible.
export async function fingerprint(passphrase: string, repoId: string): Promise<string> {
  const base = await crypto.subtle.importKey('raw', enc.encode(passphrase), 'PBKDF2', false, ['deriveBits']);
  const bits = await crypto.subtle.deriveBits(
    { name: 'PBKDF2', hash: 'SHA-256', salt: enc.encode(`cxtsecrets-fp:v1:${repoId}`) as BufferSource, iterations: ITERATIONS },
    base,
    256,
  );
  return hexstr(await crypto.subtle.digest('SHA-256', bits)).slice(0, 12);
}

export async function encryptSecrets(passphrase: string, plaintext: string, repoId: string): Promise<SecretsEnvelope> {
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const key = await deriveKey(passphrase, salt, ITERATIONS);
  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce as BufferSource, additionalData: aad(repoId) },
    key,
    enc.encode(plaintext),
  );
  return {
    version: 1,
    kdf: 'PBKDF2-SHA256',
    iterations: ITERATIONS,
    salt_b64: b64(salt),
    cipher: 'AES-256-GCM',
    nonce_b64: b64(nonce),
    ciphertext_b64: b64(ct),
    fingerprint: await fingerprint(passphrase, repoId),
  };
}

export async function decryptSecrets(
  passphrase: string,
  env: SecretsEnvelope,
  repoId: string,
  t: (k: MsgKey, v?: Vars) => string,
): Promise<string> {
  if (!validEnvelope(env)) {
    throw new Error(t('secrets.envUnsupported'));
  }
  const key = await deriveKey(passphrase, unb64(env.salt_b64), env.iterations);
  try {
    const pt = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: unb64(env.nonce_b64) as BufferSource, additionalData: aad(repoId) },
      key,
      unb64(env.ciphertext_b64) as BufferSource,
    );
    return dec.decode(pt);
  } catch {
    throw new Error(t('secrets.decryptFail'));
  }
}
